little_raft/src/state_machine.rs

```diff
@@ -63,4 +76,38 @@ where
     /// discard them. get_pending_transitions must not return the same
     /// transition twice.
     fn get_pending_transitions(&mut self) -> Vec<T>;
+
+    /// Replica calls get_snapshot once upon startup. If the Replica and the
+    /// associated StateMachine should start from a certain checkpoint
+    /// previously saved with a call to create_snapshot or set_snapshot, this
+    /// function should return Some(snapshot). Otherwise it can return None. If
+    /// None is returned, the Replica can still recover its state from other
+    /// nodes in the cluster, but it might take longer to do so than if it
+    /// recovered from a previously snapshotted value.
+    ///
+    /// Little Raft will take care of loading the Snapshot into the Replica and
+    /// achieving consensus provided snapshot.last_included_index and
+    /// snapshot.last_included_term are truthful. However, it is up to the user
+    /// to put the StateMachine into the right state before returning from
+    /// load_snapshot().
+    fn get_snapshot(&mut self) -> Option<Snapshot>;
+
+    /// create_snapshot is periodically called by the Replica if log compaction
+    /// is enabled by setting snapshot_delta > 0. The implementation MUST create
+    /// a snapshot object with truthful values of index and term.
+    ///
+    /// If the Replica should use this snapshot as a checkpoint upon restart,
+    /// the implementation MUST save the created snapshot object to permanent
+    /// storage and return it with get_snapshot after restart.
+    fn create_snapshot(
+        &mut self,
+        last_included_index: usize,
+        last_included_term: usize,
+    ) -> Snapshot;
```

Comment:
For ChiselStore, `create_snapshot()` would make an online backup of the current SQLite database to serve as an immutable snapshot:

https://www.sqlite.org/c3ref/backup_finish.html#sqlite3backupinit

However, Little Raft now stores a `Bytes` blob in `Snapshot`, which seems little cumbersome to work with. Perhaps it makes more sense to add a type parameter to `Snapshot` that allows Little Raft users to do whatever they want to represent the snapshot data?